### PR TITLE
feat(react): add jsx-curly-newline rule

### PR DESCRIPTION
**Pull request by swwind**
## Summary
- add the `react/jsx-curly-newline` rule, schema, documentation, and React catalog registration
- migrate every eslint-plugin-react v7.37.5 upstream case (excluding Flow syntax) with exact runtime source and fixer output
- add tsgo edge-case, edit-demand, and JS integration coverage

## Verification
- `go test ./internal/plugins/react/rules/jsx_curly_newline`
- `golangci-lint run --new-from-merge-base=origin/main ./internal/plugins/react/rules/jsx_curly_newline ./internal/plugins/react`
- `pnpm run check-spell <changed files>`
- `pnpm run format:check`

## Remaining
- The JS integration test could not be rerun with a rebuilt native CLI: `pnpm --filter @rslint/core build:bin` failed because the environment hit its `/tmp` build disk quota.

### CODE REVIEW

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			result.multiline = multiline
+		}
+	}
+	return result
+}
+
+func requiresNewlines(option options, expression *ast.Node, lineMap []core.TextPos, hasLeftNewline bool) bool {
+	multiline := false
+	if expression != nil {
+		multiline = expression.Pos() != expression.End() && scanner.ComputeLineOfPosition(lineMap, expression.Pos()) != scanner.ComputeLineOfPosition(lineMap, expression.End())
```
Comment (elecmonkey) [inline internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: [P1] Derive multiline status from the ESTree-equivalent expression span

This makes brace-adjacent trivia and ts-go-only wrapper nodes choose the option branch, although upstream classifies `node.expression.loc`: with `{ singleline: "forbid", multiline: "require" }`, `<App>{\nvalue}</App>` is a single-line expression upstream and is fixed to `<App>{value}</App>`, while Rslint reports `expectedBefore` and fixes it to `<App>{\nvalue\n}</App>`; parenthesized and TypeScript `as` expressions diverge similarly. The nil case also needs the authored `JSXEmptyExpression` span: upstream accepts `<App>{\n}</App>` under that option, whereas Rslint removes the newline. These ranges can be derived locally from the ts-go nodes and source text, so aligning them implements the documented distinction instead of reproducing an upstream bug.

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		position = ecmascript.SkipLeadingWhitespace(text, position, end)
+	}
+	return position
+}
+
+func skipTrailingComments(text string, position, start int) int {
+	for position > start {
+		lineStart := strings.LastIndexByte(text[start:position], '\n') + start + 1
+		if comment := strings.LastIndex(text[lineStart:position], "//"); comment >= 0 {
```
Comment (elecmonkey) [inline internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: [P2] Use parsed token and comment ranges for trailing trivia

The raw reverse searches treat token text as comments. Under `"never"`, both `<App>{'http://host'\n}</App>` and `<App>{/x*/\n}</App>` are safely autofixed upstream by removing the newline, but Rslint interprets `//` or `*/` inside the string or regexp as a trailing comment and suppresses the fix; a multiline template literal containing `http://host` can even receive a false `unexpectedBefore`. The existing parser-backed token and comment indexes distinguish these cases, so this can be fixed entirely in Rslint without a ts-go change.

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+			for position+1 < end && (text[position] != '*' || text[position+1] != '/') {
+				position++
+			}
+			if position+1 >= end {
+				return end
+			}
+			position += 2
+		case '/':
+			position += 2
+			for position < end && text[position] != '\n' && text[position] != '\r' {
```
Comment (elecmonkey) [inline internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: [P2] End line comments at every ECMAScript line terminator

This scan recognizes only CR and LF, but U+2028 and U+2029 also terminate ECMAScript line comments. For `<App>{//a\u2028value}</App>` with the default option, upstream terminates the comment before `value` and reports `expectedBefore`, while Rslint consumes `value` as comment text and reports nothing; under `"never"` it emits an extra right-side diagnostic. `ecmascript.LineTerminatorSequenceAt` or the parser's comment ranges can handle all four terminators locally, so this is not a ts-go limitation.

Comment (swwind) [reply internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: fixed all

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+				// ESLint's token APIs exclude comments. Use the parser-backed comment
+				// ranges rather than scanning text, which would mistake comment-like
+				// content in strings, regexps, and templates for trivia.
+				firstCode := skipLeadingComments(text, innerLow, innerHigh, comments)
+				lastCodeEnd := skipTrailingComments(text, innerLow, innerHigh, comments)
+				hasLeftNewline := ecmascript.ContainsLineTerminator(text, innerLow, firstCode)
+				hasRightNewline := ecmascript.ContainsLineTerminator(text, lastCodeEnd, innerHigh)
+				// ESTree expression ranges exclude brace-adjacent trivia and tsgo-only
+				// wrapper spans. Empty JSX expressions use their authored interior.
+				needsNewlines := requiresNewlines(option, firstCode, lastCodeEnd, lineMap, hasLeftNewline)
```
Comment (elecmonkey) [inline internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: [P1] Strip transparent parentheses before selecting the mode

For `<App>{(\nvalue\n)}</App>`, Espree and typescript-eslint expose the identifier itself as the expression, so it is single-line despite newlines inside transparent parentheses. Using the first and last non-comment token instead keeps those parentheses in the span: with `{ singleline: "require", multiline: "consistent" }`, upstream inserts newlines immediately inside the braces while Rslint reports nothing; with `{ singleline: "consistent", multiline: "require" }`, Rslint instead emits and applies two fixes that upstream does not. The updated 2,622-case differential matrix still has 30 diagnostic/fix mismatches, all from this remaining wrapper-span issue.

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+				if openPos < 0 || closePos <= openPos || closePos >= len(text) || text[openPos] != '{' || text[closePos] != '}' {
+					return
+				}
+
+				innerLow, innerHigh := openPos+1, closePos
+				// ESLint's token APIs exclude comments. Use the parser-backed comment
+				// ranges rather than scanning text, which would mistake comment-like
+				// content in strings, regexps, and templates for trivia.
+				firstCode := skipLeadingComments(text, innerLow, innerHigh, comments)
+				lastCodeEnd := skipTrailingComments(text, innerLow, innerHigh, comments)
```
Comment (elecmonkey) [inline internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go:93]: [P2] Avoid rescanning every file comment for each JSX expression

Both helpers iterate from the beginning or end of the complete file-level comment slice for every JSX expression, making files with many JSX containers and comments quadratic. A targeted run with 32,000 repeated comment/container pairs took about 820 ms versus 37 ms for an equivalent comment-free control, and doubling from 16,000 to 32,000 raised the commented case from about 198 ms to 820 ms. Start from the relevant comment with binary search or maintain a source-ordered cursor so each container examines only adjacent comments.

Comment (swwind) [reply internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go:93]: fixed meow

File: internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go (hunk 4 above)
Comment (swwind) [reply internal/plugins/react/rules/jsx_curly_newline/jsx_curly_newline.go]: fixed nyaa

**Review by elecmonkey — APPROVED** [review]
LGTM
